src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java

```diff
@@ -26,26 +26,75 @@
 import jdk.internal.foreign.Utils;
 import jdk.internal.foreign.abi.SharedUtils;
 
-import java.nio.ByteOrder;
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodType;
 import java.nio.charset.Charset;
 import java.util.Objects;
 import java.util.function.Consumer;
 
+import static jdk.internal.foreign.PlatformLayouts.*;
+
 /**
- * A set of utilities for working with libraries using the C language/ABI
+ * A foreign linker specializing for C Application Binary Interface (ABI) calling conventions.
+ * Instances of this interface can be used to link foreign functions in native libraries that
+ * follow the JVM's target platform C ABI.
+ *
+ * <p>There are two components that go into linking a foreign function: a method type, and
+ * a function descriptor. The method type, consisting of a set of <em>carrier</em> types constitute the
+ * set of carrier types, which, together, specify the Java signature which clients must adhere to when
+ * calling the underlying foreign function. The function descriptor contains a set of memory layouts which,
+ * together, specify the foreign function signature and classification information (via custom layout attributes),
+ * so that linking can take place. Memory layout attributes are used in the function descriptor
+ * to attach ABI classification meta-data to memory layouts, which are required for linking.
+ * Clients of this API should build function descriptors using the predefined memory layout constants
+ * (based on a subset of the built-in types provided by the C language), found in this interface;
+ * a failure to do so might result in linkage errors, given that linking requires additional classification
+ * information to determine, for instance, how arguments should be loaded into registers during a
+ * foreign function call.</p>
+ *
+ * <p>Implementations of this interface support the following primitive carrier types:
+ * {@code byte}, {@code short}, {@code char}, {@code int}, {@code long}, {@code float},
+ * and {@code double}, as well as {@link MemoryAddress} for passing pointers, and
+ * {@link MemorySegment} for passing structs and unions. Finally, the {@link VaList}
+ * carrier type can be used to match the native {@code va_list} type.</p>
+ *
+ * <p>The function descriptor used in linking contains a memory layout to match each carrier type.
+ * There are some restrictions on the carrier type and memory layout combinations that are allowed:</p>
+ * <ul>
+ *   <li>If a primitve type is used as a carrier type, the corresponding
+ *   memory layout must be a {@code ValueLayout}, and the bit size of the layout must match that of the carrier type
+ *   (see {@link Integer#SIZE} and similar fields in other primitive wrapper classes).</li>
+ *
+ *   <li>If the carrier type is {@code MemoryAddress}, then the corresponding memory layout must be a
+ *   {@code ValueLayout}, and its bit size must match the platform's address size (see {@link MemoryLayouts#ADDRESS}).
```

Comment:
super minor grammatical thing, I think "and its bit size" is more correct as "and its' bit size"

Reply:
"its" already is the possessive form, "its'"seems to be a misspelling: https://en.wiktionary.org/wiki/its%27

Reply:
Wow, sorry for the noise.